collectors/node/cpu.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2016 Mesosphere, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package node
+
+import (
+	"math"
+	"sync"
+
+	"github.com/dcos/dcos-metrics/producers"
+	"github.com/shirou/gopsutil/cpu"
+)
+
+type cpuCoresMetric struct {
+	cpuCores  int32
+	cpuTotal  float64
+	cpuUser   float64
+	cpuSystem float64
+	cpuIdle   float64
+	cpuWait   float64
+	timestamp string
+}
+
+func (m *cpuCoresMetric) poll() error {
+	ts := thisTime()
+	hi, err := getNumCores()
+	if err != nil {
+		return err
+	}
+
+	cpuStatePcts := calculatePcts(getCPUTimes())
+
+	m.cpuCores = hi
+	m.timestamp = ts
+	m.cpuTotal = cpuStatePcts.User + cpuStatePcts.System
+	m.cpuUser = cpuStatePcts.User
+	m.cpuSystem = cpuStatePcts.System
+	m.cpuIdle = cpuStatePcts.Idle
+	m.cpuWait = cpuStatePcts.Iowait
+
+	return nil
+}
+
+func (m *cpuCoresMetric) addDatapoints(nc *nodeCollector) error {
```

Comment:
do not modify the reference to a structure, return the result and let the caller handle it.
Also, this approach isn't thread safe.